library/pk_wrap.c

```diff
@@ -1095,13 +1095,92 @@ static int eckey_sign_rs_wrap(void *ctx, mbedtls_md_type_t md_alg,
 }
 #endif /* MBEDTLS_ECDSA_C && MBEDTLS_ECP_RESTARTABLE */
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+/*
+ * Alternative function used to verify that the EC private/public key pair
+ * is valid using PSA functions instead of ECP ones.
+ * The flow is:
+ * - import the private key "prv" to PSA and export its public part
+ * - write the raw content of public key "pub" to a local buffer
+ * - compare the two buffers
+ */
+static int eckey_check_pair_psa(const void *pub, const void *prv)
+{
+    psa_status_t status;
+    psa_key_attributes_t key_attr = PSA_KEY_ATTRIBUTES_INIT;
+    mbedtls_ecp_keypair *prv_ctx = (mbedtls_ecp_keypair *) prv;
+    mbedtls_ecp_keypair *pub_ctx = (mbedtls_ecp_keypair *) pub;
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    uint8_t prv_key_buf[MBEDTLS_PSA_MAX_EC_KEY_PAIR_LENGTH];
+    size_t prv_key_len;
+    uint8_t pub_key_buf[MBEDTLS_PSA_MAX_EC_PUBKEY_LENGTH];
+    size_t pub_key_len;
+    mbedtls_svc_key_id_t key_id = MBEDTLS_SVC_KEY_ID_INIT;
+    size_t curve_bits;
+    psa_ecc_family_t curve =
+        mbedtls_ecc_group_to_psa(prv_ctx->grp.id, &curve_bits);
+    size_t curve_bytes = PSA_BITS_TO_BYTES(curve_bits);
+
+    psa_set_key_type(&key_attr, PSA_KEY_TYPE_ECC_KEY_PAIR(curve));
+    psa_set_key_usage_flags(&key_attr, PSA_KEY_USAGE_EXPORT);
+
+    ret = mbedtls_mpi_write_binary(&prv_ctx->d, prv_key_buf, curve_bytes);
+    if (ret != 0) {
+        return ret;
+    }
+
+    status = psa_import_key(&key_attr, prv_key_buf, curve_bytes, &key_id);
+    if (status != PSA_SUCCESS) {
+        ret = PSA_PK_TO_MBEDTLS_ERR(status);
+        return ret;
+    }
+
+    mbedtls_platform_zeroize(prv_key_buf, sizeof(prv_key_buf));
+
+    status = psa_export_public_key(key_id, prv_key_buf, sizeof(prv_key_buf),
+                                   &prv_key_len);
+    if (status != PSA_SUCCESS) {
+        ret = PSA_PK_TO_MBEDTLS_ERR(status);
+        status = psa_destroy_key(key_id);
+        return (status != PSA_SUCCESS) ? PSA_PK_TO_MBEDTLS_ERR(status) : ret;
+    }
+
+    status = psa_destroy_key(key_id);
+    if (status != PSA_SUCCESS) {
+        return PSA_PK_TO_MBEDTLS_ERR(status);
+    }
+
+    ret = mbedtls_ecp_point_write_binary(&pub_ctx->grp, &pub_ctx->Q,
+                                         MBEDTLS_ECP_PF_UNCOMPRESSED,
+                                         &pub_key_len, pub_key_buf,
+                                         sizeof(pub_key_buf));
+    if (ret != 0) {
+        return ret;
+    }
+
+    if (memcmp(prv_key_buf, pub_key_buf, curve_bytes) != 0) {
+        return MBEDTLS_ERR_PK_BAD_INPUT_DATA;
```

Comment:
Considering this function is specific to ECC, perhaps it could return an ECP error code, for full backwards compatibility. We don't really promise to always return the exact same error codes between versions, but I think it's a bit better when it can be done easily, as is the case here.

Reply:
That's why I also had to set a "translation" of error codes in `test_suite_pk`.
I was not sure about returning ECP error codes in PK module: it is OK now that we have a sort of duality with these 2 functions, but if in the future we'll remove the ECP one then I think it will be strange to return an ECP error here. Wdyt?
BTW if we keep the ECP error code in PK then I will write a note about the reason that's behind that decision.

Reply:
Yeah, it's the change in the test that made me notice :) Generally speaking, having to change the tests is a compatibility smell as I was saying in the other PR.

We have exceptions that cover error codes, see the list [here](https://github.com/Mbed-TLS/mbedtls/blob/development/BRANCHES.md#backwards-compatibility-for-application-code), and in practice most people will just check for `!= 0` anyway. Also, in that case [the documentation](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/pk.h#L716) does not mention `ECP_BAD_INPUT_DATA` explicitly, only "Another non-zero value if the keys do not match." so I think the change is acceptable.

Also, I take your point the returning a PK error code makes more sense, especially in the future when we allow compiling out ECP.

But as a general principle, I'm always a bit happier when we don't have to change the tests at all.

Honestly, I think both ways (returning the PK or ECP code) make sense, and none of them is bad, so I won't insist. Let's see what the other reviewer thinks, and if they don't have a strong opinion about it we can keep the code as it is.